### PR TITLE
convert sparkline api to one request - fixes archive

### LINKED ISSUE #18088: Multiple requests may archive the same range request at the same time
We noticed a while ago that multiple requests may launch archiving for the same period.

To reproduce this:

* Disable browser archiving & Disable browser archiving for a segment
  * eg set `[General]enable_browser_archiving_triggering = 0
enable_general_settings_admin = 0
browser_archiving_disabled_enforce = 1`
* Have a site with say 30 active goal 
* Have generally data for a few days 
* Trigger archiving using `./console core:archive`
* Then go the Goals Overview page and request a date range eg `2021-09-02,2021,09,29`

You might then notice that all does goals trigger the archiving. Eg you could execute a query like 

```sql
select count(*) from archive_numeric_2021_09 where ts_archived > DATE_SUB(NOW(), INTERVAL 5 MINUTE) and ts_archived < NOW() and name like 'done%'; -- number of unique added archives in the last few minutes which may be pretty much roughly like the number of goals
select count(*) from archive_numeric_2021_09 where ts_archived > DATE_SUB(NOW(), INTERVAL 5 MINUTE) and ts_archived < NOW() ;// how many archive entries were added overall

-- same for blob
select count(*) from archive_blob_2021_09 where ts_archived > DATE_SUB(NOW(), INTERVAL 5 MINUTE) and ts_archived < NOW() ;
```

I'm meaning that each of these requests like in the screenshot seem to be launching the archiving (eg `index.php?forceView=1&viewDataTable=sparklines&module=Goals&action=get&idGoal=7&allow_multiple=0&only_summary=1&idSite=12&period=week&date=today&segment=&showtitle=1&random=1482`):

image

As I'm seeing this recently more often I think this was a regression somewhere between 4.3 and 4.4. I quickly looked through the archiving related changes in https://github.com/matomo-org/matomo/compare/4.3.0...4.4.1 and couldn't find anything so far that would explain this. This one peaked my interest but pretty sure it cannot be what caused it: https://github.com/matomo-org/matomo/pull/17657/files

Possibly the regression was in 4.2.0 because of https://github.com/matomo-org/matomo/pull/17379/files maybe

I wonder if this one maybe previously prevented that each request launches the archiving at the same time? Once we can reproduce the issue, we could maybe see if it works with that PR and wouldn't launch multiple archives.

Let me know if there are any issues reproducing this issue.

refs https://github.com/matomo-org/matomo/issues/17774

The goal is that the archiving is triggered only once for all these requests. Meaning below query should return `1`:

```sql
select count(*) from archive_numeric_2021_09 where ts_archived > DATE_SUB(NOW(), INTERVAL 5 MINUTE) and ts_archived < NOW() and name like 'done%'; -- number of unique added archives in the last few minutes which may be pretty much roughly like the number of goals
```

### PR DESCRIPTION
**Pull request by peterhashair**
### Description:

Description:
Fixes: #18088.

Follow the fix on UI tests

### Review

* [ ] [Functional review done](https://developer.matomo.org/guides/pull-request-reviews#functional-review-done)
* [ ] [Potential edge cases thought about](https://developer.matomo.org/guides/pull-request-reviews#potential-edge-cases-thought-about) (behavior of the code with strange input, with strange internal state or possible interactions with other Matomo subsystems)
* [ ] [Usability review done](https://developer.matomo.org/guides/pull-request-reviews#usability-review-done) (is anything maybe unclear or think about anything that would cause people to reach out to support)
* [ ] [Security review done](https://developer.matomo.org/guides/security-in-piwik#checklist)
* [ ] [Wording review done](https://developer.matomo.org/guides/pull-request-reviews#translations-wording-review-done)
* [ ] [Code review done](https://developer.matomo.org/guides/pull-request-reviews#code-review-done)
* [ ] [Tests were added if useful/possible](https://developer.matomo.org/guides/pull-request-reviews#tests-were-added-if-usefulpossible)
* [ ] [Reviewed for breaking changes](https://developer.matomo.org/guides/pull-request-reviews#reviewed-for-breaking-changes)
* [ ] [Developer changelog updated if needed](https://developer.matomo.org/guides/pull-request-reviews#developer-changelog-updated-if-needed)
* [ ] [Documentation added if needed](https://developer.matomo.org/guides/pull-request-reviews#documentation-added-if-needed)
* [ ] Existing documentation updated if needed

### CODE REVIEW

**Comment by peterhashair** [comment]
this is the follow-up on the UI failed fixes. Should pass the test now.

**Comment by bx80** [comment]
@peterhashair The only remaining failing UI screenshot `UIIntegrationTest_dashboard2.png` is expecting the goals overview widget goal titles to be wrapped in single quotes, where as the processed screenshot has them without.
image
It seems like it might be related to this change?

**Review by bx80 — APPROVED** [review]
UI tests are now all passing. 

I noticed that because the `GetMetrics` report is a now a descendant of the `Get` report that the view is configured slightly differently with overview Goal Names shown as `Goal 'Something'` instead of `Goal Something`. This would seem to be more consistent, what do you think @sgiehl ?

File: plugins/Goals/Pages.php [hunk 1]
```diff
@@ -54,9 +54,21 @@ public function createGoalsOverviewPage($goals)
...
+        $config = $this->factory->createCustomWidget('getSparklines');
+        $config->setSubcategoryId($subcategory);
+        $config->setName('');
+        $config->setOrder(25);
+        $config->setIsNotWidgetizable();
+        $widgets[] = $config;
+
+        $container = $this->createWidgetizableWidgetContainer('GoalsOverview', $subcategory, $widgets);
+
         foreach ($goals as $goal) {
```
Comment (sgiehl) [inline plugins/Goals/Pages.php]: I guess this `foreach` is no obsolete. As `$widgets` isn't used afterwards anymore, there is no sense it filling it more.

File: plugins/Goals/Controller.php [hunk 2]
```diff
@@ -219,6 +223,37 @@ public function getEvolutionGraph(array $columns = array(), $idGoal = false, arr
...
+    {
+        $module = "Goals";
+        $action = "getMetrics";
+        $content = "";
+
+        $idSite = Common::getRequestVar('idSite', null, 'int');
+        $goals = Request::processRequest('Goals.getGoals', ['idSite' => $idSite, 'filter_limit' => '-1'], []);
+
+        $apiProxy = Proxy::getInstance();
+        $queryString = $_SERVER['QUERY_STRING'];
```
Comment (sgiehl) [inline plugins/Goals/Controller.php]: seems unused
```suggestion
```

File: plugins/Goals/Controller.php [hunk 3]
```diff
@@ -219,6 +223,37 @@ public function getEvolutionGraph(array $columns = array(), $idGoal = false, arr
...
     }
 
+    public function getSparklines()
+    {
+        $module = "Goals";
+        $action = "getMetrics";
+        $content = "";
+
+        $idSite = Common::getRequestVar('idSite', null, 'int');
+        $goals = Request::processRequest('Goals.getGoals', ['idSite' => $idSite, 'filter_limit' => '-1'], []);
```
Comment (sgiehl) [inline plugins/Goals/Controller.php]:
```suggestion
        $goals = Request::processRequest('Goals.getGoals', ['idSite' => $this->idSite, 'filter_limit' => '-1'], []);
```

File: plugins/Goals/Controller.php [hunk 4]
```diff
@@ -219,6 +223,37 @@ public function getEvolutionGraph(array $columns = array(), $idGoal = false, arr
...
+
+        $apiProxy = Proxy::getInstance();
+        $queryString = $_SERVER['QUERY_STRING'];
+        if (!$apiProxy->isExistingApiAction($module, $action)) {
+            throw new \Exception("Invalid action name '$action' for '$module' plugin.");
+        }
+
+        $apiAction = $apiProxy->buildApiActionName($module, $action);
```
Comment (sgiehl) [inline plugins/Goals/Controller.php]: Can't see much value in checking and building the API action name using those methods. Guess you could also simply use `Goals.getMetrics`. If that one gets deleted somewhen, this method will throw an exception nevertheless, as the report can't be found at another point while processing the view.

**Review by sgiehl — CHANGES_REQUESTED** [review]
Generally seems to work now. Left a couple of comments for code improvements

File: plugins/Goals/Controller.php [hunk 5]
```diff
@@ -219,6 +223,25 @@ public function getEvolutionGraph(array $columns = array(), $idGoal = false, arr
...
+
+        foreach ($goals as $goal) {
+            //this is used to hack query params
+            $_GET['idGoal'] = $goal['idgoal'];
+            $_GET['allow_multiple'] = (int)$goal['allow_multiple'];
+            $_GET['only_summary'] = 1;
+            //load Visualisations Sparkline
+            $view = ViewDataTableFactory::build(Sparklines::ID, 'Goals.getMetrics', 'Goals.' . __METHOD__, true);
+            $view->config->show_title = true;
+            $content .= $view->render();
```
Comment (sgiehl) [inline plugins/Goals/Controller.php]: Here you could use `Context::executeWithQueryParameters`, which will ensure the request parameters are reset afterwards.
```suggestion
            $params = [
                'idGoal' => $goal['idgoal'],
                'allow_multiple' => (int) $goal['allow_multiple'],
                'only_summary' => 1,
            ];

            \Piwik\Context::executeWithQueryParameters($params, function() use (&$content) {
                //load Visualisations Sparkline
                $view = ViewDataTableFactory::build(Sparklines::ID, 'Goals.getMetrics', 'Goals.' . __METHOD__, true);
                $view->config->show_title = true;
                $content .= $view->render();
            });
```

**Review by sgiehl — CHANGES_REQUESTED** [review]
Had another idea for further code improvement. Otherwise this should be good to merge, if tests are passing.

**Pull request merged**